util/workloadidentity/workloadidentity.go

```diff
@@ -29,17 +35,29 @@ func NewWorkloadIdentityTokenProvider() TokenProvider {
 	return WorkloadIdentityTokenProvider{tokenCredential: cred}
 }
 
-func (c WorkloadIdentityTokenProvider) GetToken(scope string) (string, error) {
+func (c WorkloadIdentityTokenProvider) GetToken(scope string) (*Token, error) {
 	if initError != nil {
-		return "", initError
+		return nil, initError
 	}
 
 	token, err := c.tokenCredential.GetToken(context.Background(), policy.TokenRequestOptions{
 		Scopes: []string{scope},
 	})
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	return token.Token, nil
+	return &Token{AccessToken: token.Token, ExpiresOn: token.ExpiresOn}, nil
+}
+
+func CalclulateCacheExpiryBasedOnTokenExpiry(tokenExpiry time.Time) time.Duration {
+	// Calculate the cache expiry as 5 minutes before the token expires
+	cacheExpiry := time.Until(tokenExpiry) - time.Minute*5
+
+	// If token exires in negative time, cache will never expire
+	// If expiry is 0 then default cache expiry is used.
+	if cacheExpiry <= 0 {
+		cacheExpiry = time.Second
+	}
```

Comment:
The comment doesn't seem to match the behavior. If the time until the token expiry (minus 5 minutes) is less than or equal to zero, then this code sets the expiry to one second, not never.

Reply:
Done